client-side-virtual-keyboard.md

```diff
@@ -24,87 +23,46 @@ After importing the AAR file, use the AnboxWebView in an Android project:
+## Build communication bridge
 
+In your HTML code, define the JavaScript functions that can be invoked from the Android Java layer (AnboxWebView). Those functions are just wrappers of the APIs that are exposed from the JavaScript SDK. Defining them in your code builds up a communication tunnel between the Android Java layer (AnboxWebView) and the JavaScript layer.
    ```
-   import com.canonical.anbox.streaming_sdk.AnboxWebView;
-   ...
-
-   public class MainActivity extends AppCompatActivity {
-       protected void onCreate(Bundle savedInstanceState) {
-           ...
-           mWebview = (AnboxWebView) findViewById(R.id.webview);
-           mWebView.addJavascriptInterface(new AppInterface(this), "AppInterface");
-           ...
-       }
-   }
-   ```
-
-3. For the application running in the Android container, whenever a text editor gains or loses focus during streaming, an IME-specific message is sent from the Android container to the client through the JavaScript SDK. Each message requests a certain IME operation. This allows you to perform one certain action when the message is received and ensures the behaviour of the virtual keyboard is synced on both ends. The supported IME messages are as follows:
-
-   The following IME messages are supported:
-
-    Message          |  Description
-    -----------------|-----------------------------------------------
-    show-ime         |  Request to show client-side virtual keyboard
-    hide-ime         |  Request to hide client-side virtual keyboard
-
-    To receive the above messages, register the `messageReceived` callback function when instantiating an AnboxStream object in HTML. Through the injected Java object `AppInterface`, the client-side virtual keyboard will pop up or hide when the corresponding message is received.
-
-   ```
-       window.stream = new AnboxStream({
-           ...
-           ...
-           ...
-           callbacks: {
-               ...
-               ...
-               messageReceived: (type, data) => {
-                   switch(type) {
-                   case "show-ime" :
-                       AppInterface.openVirtualKeyboard();
-                   break;
-                   case "hide-ime" :
-                       AppInterface.hideVirtualKeyboard();
-                   break;
-                   default:
-                       console.log("Unhandled message:" + type)
-                   break;
-                   }
-               }
-           }
-       });
-   ```
-
-4. Set up the listener to the activity where you want to capture the message. You must implement the following methods for the AppInterface.ActionListener interfaces so that the application will be notified when a request is coming from the JavaScript SDK.
+   <script>
+     function sendIMECommittedText(text) {
+       window.stream.sendIMECommittedText(text);
+     }
+     function sendIMEComposingText(text) {
```

Comment:
This is a bit confusing. As reader, I have a few questions:
1. What are those functions ? You mention `define the JavaScript functions that can be invoked from the Android Java layer` but it's not specified whether the user has to copy paste the given code snippet, or if he has to implement something.
2. Where should these functions be defined ? What/who calls those functions ? Some more complex web apps might not want to expose those functions in a `<script></script>` tag

Reply:
What are those functions ? You mention define the JavaScript functions that can be invoked from the Android Java layer but it's not specified whether the user has to copy paste the given code snippet, or if he has to implement something.
A: User doesn't need to implement actual logic how IME text is sent from Android java layer to JavaScript, as the caller, they just need to define those functions, which acts as a bridge for the communication between Android application(more specifically com.android.) and JS SDK through the HTML code.
2. Where should these functions be defined ? 
A: From the context, those functions must be defined in your HTML(JS) code, 
4. What/who calls those functions ? 
A: AnboxWebView component which is provided from `com.canonical.anbox.streaming_sdk.aar` is going to call those functions. That means for most common case, in the actual java code, developers just need to import the `com.canonical.anbox.streaming_sdk.aar` file and use AnboxWebView in their layout XML file. That's it. And they don't need to care about the protocol that is used on both ends. We will help them to figure out that, that's the major change(refractor) we made in 1.11.2. I'd suggest taking a look at this change and commit message for the reason behind the sense.
https://code.launchpad.net/~gary-wzl77/indore/+git/anbox-streaming-stack/+merge/407715
6. Some more complex web apps might not want to expose those functions in a <script></script> tag
A: Web developers don't need to implement those functions if they don't want to integrate client-side virtual keyboard. That's being said, if you want to integrate the client side virtual keyboard, following the interface definition, provide those functions in your HTML(JS) code, everything should be working under the hood. If you don't want to expose those functions, there is no way for the Android java layer to talk to the js sdk to communicate with Android container through the tunnel

Reply:
**they just need to define those functions**
I see, so it's just a matter of wording. Instead of 
`define the JavaScript functions that can be invoked from the Android Java layer`
we could have something like 
`copy the following Javascript functions in your web application. They will automatically be invoked by the Android java layer.`

**provide those functions in your HTML(JS) code**
That's the thing I was getting at, we should specify that those functions must be attached to the global `window` object. For example, if a user defines those function elsewhere, e.g. `window.someFramework.views.main.anboxIntegration` then this would not work

Reply:
how about
```
Adding the following JavaScript code snippet builds up a communication tunnel between the Android Java layer (AnboxWebView) and the JavaScript layer. It attaches the AnboxStream object to the global `window` variable. And those functions will automatically be invoked by the Android java layer(AnboxWebView).
```